### PR TITLE
Fix C# native library loading in .NET 5 single-file apps

### PR DESCRIPTION
**Pull request by jtattermusch**
Fixes https://github.com/grpc/grpc/issues/24266.

### CODE REVIEW

**Comment by jtattermusch** [comment]
Once this is approved, I'll also create a backport PR.

**Comment by jtattermusch** [comment]
Adhoc distribtests: https://g3c.corp.google.com/results/invocations/a25cde5a-bf1c-49cc-923e-315e96987e17 (including the newly added single file publish distribtest).

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs [hunk 1]
```diff
@@ -153,13 +153,20 @@ private static NativeMethods LoadNativeMethodsXamarin()
...
-        private static string GetAssemblyPath()
+        private static string GetAssemblyDirectory()
         {
             var assembly = typeof(NativeExtension).GetTypeInfo().Assembly;
 #if NETSTANDARD
             // Assembly.EscapedCodeBase does not exist under CoreCLR, but assemblies imported from a nuget package
             // don't seem to be shadowed by DNX-based projects at all.
-            return assembly.Location;
+            var assemblyLocation = assembly.Location;
+            if (assemblyLocation != null)
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs:163]: Actually, assembly.Location might be returning an empty string according to the docs:
https://docs.microsoft.com/en-us/dotnet/core/deploying/single-file#api-incompatibility

I'll check which one is correct and update the code if needed.

**Comment by jtattermusch** [comment]
Another adhoc distribtest run: https://g3c.corp.google.com/results/invocations/8872dc43-c7da-4533-b792-bc1c5eaff03b/targets

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs [hunk 2]
```diff
@@ -153,13 +153,20 @@ private static NativeMethods LoadNativeMethodsXamarin()
...
-        private static string GetAssemblyPath()
+        private static string GetAssemblyDirectory()
         {
             var assembly = typeof(NativeExtension).GetTypeInfo().Assembly;
 #if NETSTANDARD
             // Assembly.EscapedCodeBase does not exist under CoreCLR, but assemblies imported from a nuget package
             // don't seem to be shadowed by DNX-based projects at all.
-            return assembly.Location;
+            var assemblyLocation = assembly.Location;
+            if (string.IsNullOrEmpty(assemblyLocation))
```
Comment (jkotas) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs:163]:
```suggestion
            if (!string.IsNullOrEmpty(assemblyLocation))
```

**Comment by jtattermusch** [comment]
Yet another adhoc distribtest: https://g3c.corp.google.com/results/invocations/700ebfdb-8e9a-43c8-9e43-2e58a4929037

**Comment by jkotas** [comment]
https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core/Internal/UnmanagedLibrary.cs#L245 is another problem that needs to be fixed to make the single file scenario working. There is no libcoreclr.so on disk in the single file scenario.

**Comment by jtattermusch** [comment]
> https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core/Internal/UnmanagedLibrary.cs#L245 is another problem that needs to be fixed to make the single file scenario working. There is no libcoreclr.so on disk in the single file scenario.

Ok, thanks for the tip. Do you have ideas on how to fix this problem?

**Comment by gitfool** [comment]
@jtattermusch please add a distribtest for the `IncludeNativeLibrariesForSelfExtract` case. Hopefully @jkotas can advise on how to detect and/or handle that too?

**Comment by jkotas** [comment]
> Do you have ideas on how to fix this problem?

The best way to fix this problem is to use [`Native.Load`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary.load?#System_Runtime_InteropServices_NativeLibrary_Load_System_String_System_Reflection_Assembly_System_Nullable_System_Runtime_InteropServices_DllImportSearchPath__) for .NET Core. This API deals with all probing paths, single file and platform differences.

**Review by apolcyn — APPROVED** [review]

**Comment by jtattermusch** [comment]
> Native.Load

Thanks for the tip! Unfortunately Native.Load is only available on .NET Core 3+ (and at this point the newest we target is netstandard2.0)

**Comment by jtattermusch** [comment]
Another distribtest: https://g3c.corp.google.com/results/invocations/757851c5-c719-4434-bad6-bac0e655990b/targets

**Comment by jtattermusch** [comment]
The interop test failure is https://github.com/grpc/grpc/issues/24764.

**Comment by jtattermusch** [comment]
As @jkotas  has pointed out, I'm now getting the `Unhandled exception. System.DllNotFoundException: Unable to load shared library 'libcoreclr.so'` error.

https://source.cloud.google.com/results/invocations/9ccd7ae8-32c0-4ee9-b08a-e7848b2f84a4/targets/github%2Fgrpc/tests

**Comment by jtattermusch** [comment]
@yulin-liang  I don't have a fully working fix at this point, so I think it's fine to cut the release branch without it. (If I come up with a fix, I can backport it)

**Comment by jkotas** [comment]
> Unfortunately Native.Load is only available on .NET Core 3+ (and at this point the newest we target is netstandard2.0)

The fix for this should be to try to lookup these methods using reflection, and then use them instead of the hardcoded P/Invokes when these methods are available. Something like:

```csharp
delegate bool TryLoadDelegate(string libraryName, System.Reflection.Assembly assembly, System.Runtime.InteropServices.DllImportSearchPath? searchPath, out IntPtr handle);
delegate bool TryGetExportDelegate(IntPtr handle, string name, out IntPtr address);

static TryLoadDelegate s_tryLoad;
static TryGetExportDelegate s_tryGetExport; 

...

Type nativeLibrary = Type.GetType("System.Runtime.InteropServices.NativeLibrary, System.Runtime.InteropServices, PublicKeyToken=b03f5f7f11d50a3a", throwOnError: false);
if (nativeLibrary == null)
    return;

MethodInfo tryLoad = nativeLibrary.GetMethod("TryLoad", BindingFlags.Static | BindingFlags.Public, null,
    new Type[] { typeof(string), typeof(Assembly), typeof(Nullable<DllImportSearchPath>), typeof(IntPtr).MakeByRefType() }, null);
s_tryLoad = (TryLoadDelegate)tryLoad?.CreateDelegate(typeof(TryLoadDelegate));

MethodInfo tryGetExport = nativeLibrary.GetMethod("TryLoad", BindingFlags.Static | BindingFlags.Public, null,
    new Type[] { typeof(IntPtr), typeof(string), typeof(IntPtr).MakeByRefType() }, null);
s_tryGetExport = (TryGetExportDelegate)tryGetExport?.CreateDelegate(typeof(TryGetExportDelegate));
```

**Comment by jtattermusch** [comment]
Alternative approach used by a library that faced a similar problem: https://github.com/migueldeicaza/gui.cs/commit/19955faa8beb435d37007de62f7102803fbecde9#diff-91d94d0e5b8dd01a182c51299d7e8093247fa636a9768574b7042bfd5df9a1deR261

**Comment by jtattermusch** [comment]
I refactored the native library loading logic to use the default `DllImport` loading mechanism when on .NET Core.
It leads to some code simplification and also seems to handle the case with of .NET 5 single file published app well.
PTAL.

The adhoc distribtest run is here: https://g3c.corp.google.com/results/invocations/4da3122d-8972-48ff-ac92-d01b22705b99/log

**Comment by jtattermusch** [comment]
Known failures: build_docker_go is also failing on master.

**Comment by jtattermusch** [comment]
Adhoc distribtests are passing:
https://g3c.corp.google.com/results/invocations/50ae873b-5510-4ae0-b735-0840c7e9dd6d/targets/grpc%2Fcore%2Fexperimental%2Fgrpc_distribtests_multiplatform/log (the php7 distribtest has actually failed but that's unrelated to this PR - everything C# related is green)

dotnet 5 distribtest is also green:
https://g3c.corp.google.com/results/invocations/4228feb1-e59b-4bb1-9933-982998e8d431/targets/github%2Fgrpc;config=default/tests;query=csharp

**Comment by jtattermusch** [comment]
@jkotas  your opinions on the approach I ended up taking (= rely on `DllImport` default loading logic) are welcome.

@apolcyn  PTAL.

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs [hunk 3]
```diff
@@ -80,28 +80,25 @@ public NativeMethods NativeMethods
         /// <summary>
         /// Detects which configuration of native extension to load and load it.
         /// </summary>
-        private static UnmanagedLibrary LoadUnmanagedLibrary()
+        private static NativeMethods LoadNativeMethodsLegacyNetFramework()
         {
             // TODO: allow customizing path to native extension (possibly through exposing a GrpcEnvironment property).
             // See https://github.com/grpc/grpc/pull/7303 for one option.
-            var assemblyDirectory = Path.GetDirectoryName(GetAssemblyPath());
+            var assemblyDirectory = GetAssemblyDirectory();
```
Comment (jkotas) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs:87]: The changes in this method should not be needed with your new fix.

File: src/csharp/Grpc.Core/Internal/NativeExtension.cs [hunk 4]
```diff
@@ -117,7 +114,43 @@ private static NativeMethods LoadNativeMethods()
...
+                // While in theory, we could just [DllImport("grpc_csharp_ext")] for all the platforms
+                // and operating systems, the native libraries in the nuget package
+                // need to be laid out in a way that still allows things to work well under
+                // the legacy .NET Framework (where native libraries are a concept unknown to the runtime).
+                // Therefore, we use several flavors of the DllImport attribute
+                // (e.g. the ".x86" vs ".x64" suffix) and we choose the one we want at runtime.
+                // The classes with the list of DllImport'd methods are code generated,
+                // so having more than just one doesn't really bother us.
+
+                // on Windows, the DllImport("grpc_csharp_ext.x64") doesn't work for some reason,
```
Comment (jkotas) [inline src/csharp/Grpc.Core/Internal/NativeExtension.cs:133]: This was bug https://github.com/dotnet/coreclr/pull/17505 . It is fixed in .NET Core 3.1, but I assume you still want this to work on .NET Core 2.1.

**Comment by jkotas** [comment]
> @jkotas your opinions on the approach I ended up taking

Looks fine to me.

**Review by apolcyn — APPROVED** [review]

**Comment by gitfool** [comment]
@jtattermusch does your fix handle the `IncludeNativeLibrariesForSelfExtract` case too?

**Comment by jtattermusch** [comment]
> @jtattermusch does your fix handle the `IncludeNativeLibrariesForSelfExtract` case too?

I haven't tested that explicitly - would you mind trying it out and posting back the result?

The nightly build of the nugets should appear here soon: 
https://packages.grpc.io/

Alternatively, here are nugets from my last manual build:
https://storage.cloud.google.com/grpc-testing-kokoro-prod/test_result_public/prod/grpc/core/master/linux/grpc_build_packages/14907/20201130-052042/github/grpc/artifacts/csharp_nugets_windows_dotnetcli.zip

I wanted to merge this PR ASAP because like this the fixes will still make it to Grpc.Core 2.34.0 release - we can address `IncludeNativeLibrariesForSelfExtract` as a follow up.

**Comment by gitfool** [comment]
@jtattermusch I did a test with `Grpc.Core` version `2.34.0` and it works for `IncludeNativeLibrariesForSelfExtract` too! 😀

**Pull request merged**